### PR TITLE
Bug 1907421: Implementing retries during for image mirror

### PR DESCRIPTION
**Pull request by ricardomaraschini**
This commits adds a retry logic when copying a blob from one registry to another in case of "REFUSED_STREAM" error. I have spent a considerable amount of time trying to replicate this behavior together with @hongkailiu without success.

We could see the REFUSED_STREAM error a few times when running the `oc image mirror` command using an image based on the master branch of this repository but we were not so lucky when it comes to see the same error using an image based on this very same branch. 

If it is worthy I can assure the command still works after these changes but I can not assure this will fix the issue, if we manage to get this in we can validate if it actually works by examining the output for some [jobs](https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/branch-ci-openshift-ci-tools-master-images/1352335725302910976), we should not see this failure anymore.

### CODE REVIEW

**Comment by hongkailiu** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_oc/706/pull-ci-openshift-oc-master-images/1349713704706379776

```bash
oc --context build02 tag   ci-op-3i82xw33/pipeline@sha256:f1d5a5e0ccea6e832fca401e2a69123bf197c2cd3225cb77294c4aa7e0d1b8f0  hongkliu-test/cli:b00f05f
Tag cli:b00f05f set to ci-op-3i82xw33/pipeline@sha256:f1d5a5e0ccea6e832fca401e2a69123bf197c2cd3225cb77294c4aa7e0d1b8f0.
```

File: pkg/cli/image/mirror/mirror.go [hunk 1]
```diff
@@ -743,11 +744,21 @@ func copyBlob(ctx context.Context, plan *workPlan, c *repositoryBlobCopy, blob d
...
-		return err
 	}
-	return nil
+
+	var err error
+	for i := 0; i < 5; i++ {
+		if err = copyfn(); err != nil {
+			if strings.Contains(err.Error(), "REFUSED_STREAM") {
+				klog.V(6).Infof("retrying push due to %s", err.Error())
+				time.Sleep(time.Second)
```
Comment (soltysh) [inline pkg/cli/image/mirror/mirror.go:754]: How about using `k8s.io/client-go/util/retry/util#OnError` or at least `k8s.io/client-go/util/flowcontrol#Backoff`?

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/cli/image/mirror/mirror.go (hunk 1 above)
Comment (ricardomaraschini) [reply pkg/cli/image/mirror/mirror.go]: Yes, that would be neat. Done.

File: pkg/cli/image/mirror/mirror.go [hunk 2]
```diff
@@ -745,11 +748,23 @@ func copyBlob(ctx context.Context, plan *workPlan, c *repositoryBlobCopy, blob d
...
-	return nil
+
+	return retry.OnError(
+		wait.Backoff{
+			Duration: time.Second,
+			Factor:   1,
+			Steps:    5,
+		},
+		func(err error) bool {
+			if strings.Contains(err.Error(), "REFUSED_STREAM") {
```
Comment (soltysh) [inline pkg/cli/image/mirror/mirror.go:760]: `return strings.Contains(err.Error(), "REFUSED_STREAM")`? It seems simpler and you can always log in `copyFn`, especially that the error is expected one.

Comment (ricardomaraschini) [reply pkg/cli/image/mirror/mirror.go:760]: Yeah, I am planning to remove this log - as soon as I manage to replicate the issue with `master`. I will get back to you.

Comment (dmage) [reply pkg/cli/image/mirror/mirror.go]: Will it be possible to understand that `oc` hit the error (with debug logging enabled)?

Comment (ricardomaraschini) [reply pkg/cli/image/mirror/mirror.go]: No, I have removed the log. Do you want me to put it back?

Comment (ricardomaraschini) [reply pkg/cli/image/mirror/mirror.go]: I changed my mind, there is value in having this logged. Following @soltysh idea I have chosen to log in the function body instead, PTAL.

https://github.com/openshift/oc/pull/706/files#diff-1e91260f74d6221c0e5d1b5fcf70982d26d168e294254fea0b6baf8b771f7bcaR741

File: pkg/cli/image/mirror/mirror.go [hunk 3]
```diff
@@ -745,11 +749,19 @@ func copyBlob(ctx context.Context, plan *workPlan, c *repositoryBlobCopy, blob d
...
 		plan.BytesCopied(n)
 		return nil
-	}()
-	if err != nil {
-		return err
 	}
-	return nil
+
+	return retry.OnError(
+		wait.Backoff{
```
Comment (soltysh) [inline pkg/cli/image/mirror/mirror.go:755]: Final nit: `retry.DefaultRetry`

**Review by soltysh — APPROVED** [review]
/approve

File: pkg/cli/image/mirror/mirror.go (hunk 3 above)
Comment (ricardomaraschini) [reply pkg/cli/image/mirror/mirror.go]: done.

**Review by soltysh — APPROVED** [review]
/lgtm
/retest

**Comment by smarterclayton** [comment]
I don't understand this change.  Retrying should be a function of the lower level image connection, not a high level component like mirror.

If you believe this is a retriable issue, it should be a special case in library-go, not in mirror.  I'd like this reverted and moved to library-go and I'll do a review there.

**Comment by soltysh** [comment]
@ricardomaraschini can you handle this ^

**Comment by ricardomaraschini** [comment]
@smarterclayton I agree that this PR looks tricky. Let me do a quick recap on its history and hopefully you may be able to add something I may not be not seeing:

1. REFUSED_STREAM is a retriable error.
2. This error happens during `ReadFrom` where we are essentially piping data from one TCP connection into another.
3. Go HTTP client can auto recover from this type of error if the provided `http.Request` is capable of "getting a fresh version of the body again" (`http.Request.GetBody`).
4. When the error occurs we essentially loss the bytes that we have already read from the source connection (the source connection is the body).
5. I have tried to provide the request with a `GetBody` by caching the layer into a local file (https://github.com/openshift/oc/compare/master...ricardomaraschini:cache-to-local-file), solution that was at the time considered not adequate so I haven't further developed the idea.
6. Hence this PR.

As we can't simply "provide a new version of the body" without caching it somehow and the ownership of the "source tcp connection" belongs to the caller this solution ended up being in here and not part of the upstream "docker/distribution" repository.

With all of that being said, is there something I am not seeing?  The reasoning I had was: the ownership over the "source tcp connection" used during a `ReadFrom` operation belongs to the caller and not to the library (to cache the blob locally was deemed not adequate at the time).

/cc @soltysh @dmage

**Pull request merged**